crates/subspace-farmer/src/multi_farming.rs

```diff
@@ -0,0 +1,146 @@
+use std::{path::Path, sync::Arc, time::Duration};
+
+use anyhow::anyhow;
+use futures::stream::{FuturesUnordered, StreamExt};
+use log::info;
+use subspace_core_primitives::PublicKey;
+use subspace_solving::SubspaceCodec;
+
+use crate::{
+    Commitments, FarmerData, Farming, Identity, ObjectMappings, Plot, Plotting, RpcClient, WsRpc,
+};
+
+/// Abstraction around having multiple plots, farmings and plottings
+pub struct MultiFarming {
+    pub plots: Arc<Vec<Plot>>,
+    farmings: Vec<Farming>,
+    plottings: Vec<Plotting>,
+}
+
+impl MultiFarming {
+    /// Starts multiple farmers with any plot sizes which user gives
+    pub async fn new(
+        base_directory: impl AsRef<Path>,
+        client: WsRpc,
+        object_mappings: ObjectMappings,
+        plot_sizes: Vec<u64>,
+        reward_address: PublicKey,
+        best_block_number_check_interval: Duration,
+    ) -> anyhow::Result<Self> {
+        let mut plots = Vec::with_capacity(plot_sizes.len());
+        let mut farmings = Vec::with_capacity(plot_sizes.len());
+        let mut plottings = Vec::with_capacity(plot_sizes.len());
+
+        for (plot_index, max_plot_pieces) in plot_sizes.into_iter().enumerate() {
+            let base_directory = base_directory.as_ref().join(format!("plot{plot_index}"));
+            std::fs::create_dir_all(&base_directory)?;
+            let (plot, plotting, farming) = farm_single_plot(
+                base_directory,
+                reward_address,
+                client.clone(),
+                object_mappings.clone(),
+                max_plot_pieces,
+                best_block_number_check_interval,
+            )
+            .await?;
+            plots.push(plot);
+            farmings.push(farming);
+            plottings.push(plotting);
+        }
+
+        Ok(Self {
+            plots: Arc::new(plots),
+            farmings,
+            plottings,
+        })
+    }
+
+    /// Waits for farming and plotting completion (or errors)
+    pub async fn wait(self) -> anyhow::Result<()> {
+        let mut farming_plotting = self
+            .farmings
+            .into_iter()
+            .zip(self.plottings)
+            .into_iter()
+            .map(|(farming, plotting)| async move {
+                tokio::select! {
+                    res = plotting.wait() => if let Err(error) = res {
+                        return Err(anyhow!(error))
+                    },
+                    res = farming.wait() => if let Err(error) = res {
+                        return Err(anyhow!(error))
+                    },
+                }
+                Ok(())
+            })
+            .collect::<FuturesUnordered<_>>();
+
+        while let Some(res) = farming_plotting.next().await {
+            res?;
+        }
```

Comment:
Now I thought more about it, it should probably exit if at least one `farming_plotting` exited:
```suggestion
        if let Some(res) = farming_plotting.next().await {
            res?;
        }
```

Reply:
Should it be okay if some farmer exited with the okay and others are still alive?

Reply:
Well, if one exited there is a problem, so IMO better to stop the app in that case. We can revisit and do something like per-plot internal restarts in the future, but I don't think it'll be necessary.